spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/DVIterator.java

```diff
@@ -0,0 +1,128 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.source;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.data.BaseDeleteLoader;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.PartitionUtil;
+import org.apache.iceberg.util.StructLikeUtil;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.unsafe.types.UTF8String;
+
+class DVIterator implements CloseableIterator<InternalRow> {
+  private final PartitionSpec spec;
+  private final DeleteFile deleteFile;
+  private final Schema projection;
+  private final Map<Integer, ?> idToConstant;
+  private final Iterator<Long> positions;
+  private Integer deletedPositionIndex;
+  private GenericInternalRow row;
+
+  DVIterator(
+      InputFile inputFile,
+      DeleteFile deleteFile,
+      PartitionSpec spec,
+      Schema projection,
+      Map<Integer, ?> idToConstant) {
+    this.deleteFile = deleteFile;
+    this.spec = spec;
+    this.projection = projection;
+    this.idToConstant = idToConstant;
+    List<Long> pos = Lists.newArrayList();
+    new BaseDeleteLoader(ignored -> inputFile)
+        .loadPositionDeletes(ImmutableList.of(deleteFile), deleteFile.referencedDataFile())
+        .forEach(pos::add);
+    this.positions = pos.iterator();
+  }
+
+  @Override
+  public boolean hasNext() {
+    return positions.hasNext();
+  }
+
+  @Override
+  public InternalRow next() {
+    long position = positions.next();
+
+    if (null == row) {
+      List<Object> rowValues = Lists.newArrayList();
+      if (null != projection.findField(MetadataColumns.DELETE_FILE_PATH.fieldId())) {
+        rowValues.add(UTF8String.fromString(deleteFile.referencedDataFile()));
+      }
+
+      if (null != projection.findField(MetadataColumns.DELETE_FILE_POS.fieldId())) {
+        rowValues.add(position);
+        // remember the index where the deleted position needs to be set
+        deletedPositionIndex = rowValues.size() - 1;
+      }
+
+      Types.NestedField partition = projection.findField(MetadataColumns.PARTITION_COLUMN_ID);
+      if (null != partition) {
+        Object constant = idToConstant.get(MetadataColumns.PARTITION_COLUMN_ID);
+        if (null != constant) {
+          rowValues.add(constant);
+        } else {
+          Types.StructType type = partition.type().asStructType();
```

Comment:
Can this actually happen in practice? Shouldn't it always be populated in the constants map?

Reply:
technically it can't happen in practice. My main motivation for keeping the `else` part was to still produce valid results in case `PartitionUtil#constantsMap` ever gets changed without updating this part of the code. But I guess in such a scenario the tests would fail anyways, so we can just rely on file_path/pos/partition to always be populated. I've removed the fallback code